### PR TITLE
feat(workspace): browser-based workspace creation handoff

### PR DESCRIPTION
**Pull request by sahrizvi**
## Summary

Adds a browser-based workspace creation handoff to the CLI: post-scan / `altimate-code link` opens the Altimate SaaS on `<tenant>.ws.myaltimate.com/create-and-link` with the current project's context, the user approves in a single modal, and the SaaS delivers the newly-created workspace's id back to a CLI-local loopback listener (same pattern as gateway sign-in). CLI then binds the current project via the existing `POST /datamate-project-bindings/bind`.

**Stacked on the Workspaces draft PR (#1099 / `feat/agent-workspaces`).**

## What's added

- **`packages/opencode/src/altimate/workspace/browser-handoff.ts`** (new) — per-flow loopback listener (own instance, walks 7317..7325 past a live OAuth listener), tenant-mismatch guard, typed failure reasons. Duplicates the loopback pattern from `altimate.ts` deliberately — shared-helper refactor is a follow-up once both flows have prod experience.
- **`OfferDialog` (post-scan)** — adds "Set up in browser (recommended)" as the default option **when the deployment supports it** (freemium only for pilot; `resolveWorkspaceWebUrl` returns `null` otherwise and the option auto-hides).
- **`altimate-code link` picker** — adds "＋ Set up in browser" as the first row under the same condition.
- **Dev escape hatch** — `ALTIMATE_WORKSPACE_WEB_URL` env var overrides the deployment map lookup (used for local integration testing; never set in production).

## What's unchanged

Every pre-existing option in the post-scan dialog and `altimate-code link` picker (Create quick workspace here, Link to an existing workspace, Skip for now, existing workspace-picker rows) continues to work exactly as it does today. The browser-handoff option is **strictly additive**. Rolling back is a single-commit revert with no schema, no cache format, and no backend-contract implications.

A user whose deployment doesn't support the browser flow (localhost, enterprise) sees zero behavior change — the new option auto-hides.

## Tests

- 14 new unit tests for `browser-handoff.ts` covering URL resolution edge cases (freemium / localhost / enterprise / malformed), pre-flight failures (unavailable / not-configured), end-to-end via dependency-injected browser opener (happy path, tenant mismatch, cancel via `?error=cancelled`, missing workspace_id, invalid workspace_id, browser-open failure with `authorizeUrl` copyable), and port walk past a squatting listener on 7317.
- 32/32 workspace + plugin tests pass — no regressions.
- Typecheck clean.

## E2E verified against the live backend

Ran the CLI-side round-trip against a live `altimate-backend` on localhost:
- ✅ CLI URL construction with all expected params (`client`, `redirect`, `state`, `project_path`, `project_name`, `#cli_context`)
- ✅ Loopback listener binds + receives callback
- ✅ State + tenant validation
- ✅ Real backend `POST /datamates/` creates workspace
- ✅ Real backend `POST /bind` links with `project_path`-based binding (no git remote)
- ✅ Real backend `GET /by-path` returns the binding after
- ✅ Local cache persists to XDG state dir with correct schema
- ✅ 409-conflict handling on repeated runs is the intended behavior

Full SaaS-side E2E (real browser clicking Approve) is a manual smoke once the paired SaaS PR is up.

## Paired SaaS PR

AltimateAI/altimate-frontend PR (`feat/AI-8510-workspace-browser-handoff`) — stacked on Ralph's `feature/AI-8496-ws-list-create` branch.

## Follow-ups (deliberately out of scope)

- Loopback-listener extraction into a shared helper (currently duplicated from `altimate.ts` with a TODO comment).
- Google-first-time-on-`ws.` UX (session gate returns user to default post-auth destination instead of `/create-and-link`; password login works correctly).
- Enterprise workspace-web-host mapping.
- Removing the CLI-direct create path (a discussion for after this ships).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_016H42Vt4pt5dcD7opRqckeM

---
## Summary by cubic
Adds a browser-based workspace create-and-link flow for CLI and TUI. Previously linking ran only in the CLI; now users approve in the SaaS, the CLI receives a 127.0.0.1 callback, re-verifies credentials, and binds the project. Also adds a persistent “linked” confirmation and a right‑pane sidebar tile.

- Loopback handoff (new packages/opencode/src/altimate/workspace/browser-handoff.ts): ports 7317–7325 with Host header guard; CSRF state; tenant DNS‑label guard; strictly decimal and Number.isSafeInteger workspace_id; fragment‑only project_remote/project_path; 15‑min unref’d timeout; AbortSignal support with silent return when a newer handoff supersedes an older one; post‑listen error handler; force close of open sockets; typed failures (incl. browser_open_failed with copyable URL); success/cancel via top‑level navigation; returns a credential fingerprint for bind‑time re‑verify.
- CLI (packages/opencode/src/cli/cmd/link.ts): offers “＋ Set up in browser” only when available, unlinked, and pre‑check passes; on success binds, writes a canonical cache key (identifier.projectPath ?? directory), prints the manage URL; maps typed failures; re‑verifies credentials before binding.
- TUI (packages/opencode/src/plugin/tui/altimate/workspace.tsx): adds “Set up in browser (recommended)” when available; guards credential errors; aborts stale handoffs via a module‑level AbortController and silently ignores superseded results; re‑verifies credentials before bind; replaces success toasts with a persistent linked dialog that can open a safe http(s) manage URL.
- Sidebar (packages/opencode/src/plugin/tui/altimate/workspace-sidebar.tsx, registered in packages/opencode/src/plugin/tui/altimate/index.ts): read‑only tile shows current workspace and manage URL, marks “(pinned via --workspace)” when applicable, polls every 30s with memoized manage base; gated by the same flag.
- Local cache (packages/opencode/src/altimate/workspace/state.ts): canonicalizes directory keys via realpathSync on read/write with a one‑shot migration; wraps migration writes in try/catch; best‑effort writes.
- Availability: resolveWorkspaceWebUrl maps freemium only (api.myaltimate.com → <tenant>.ws.myaltimate.com), validates tenants, and honors ALTIMATE_WORKSPACE_WEB_URL for dev.
- Tests (packages/opencode/test/altimate/workspace/browser-handoff.test.ts): cover URL resolution, preflight errors, tenant mismatch, cancel, CSRF state‑mismatch where the legitimate callback wins, fragment handling, port walk, browser‑open failure; isolate ALTIMATE_WORKSPACE_WEB_URL.

- Rollout
  - Behind Flag.ALTIMATE_WORKSPACE (TUI plugin, sidebar, post‑scan offer, CLI menu).
  - Freemium only; enterprise/local/custom domains hide the browser option.
  - No backend migrations; local cache migration is transparent. Existing CLI link flows remain unchanged.

<sup>Written for commit 85161a5824b2c30e701400c6ef79c3782c9e3cc7. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/AltimateAI/altimate-code/pull/1100?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

## Summary by CodeRabbit

* **New Features**
  * Added workspace linking and management through the `link` command.
  * Added browser-based workspace setup, direct creation, and existing workspace selection.
  * Added workspace status and management details to the TUI sidebar.
  * Added optional workspace tools controlled by a feature flag.
  * Added local binding persistence for faster workspace access.
* **Bug Fixes**
  * Added validation for credentials, tenants, callbacks, permissions, and unsafe URLs.
  * Added handling for offline use, cancellations, timeouts, and setup failures.
* **Tests**
  * Added coverage for browser handoff, callback validation, URL handling, and fallback behavior.

### CODE REVIEW

**Review by saravmajestic — DISMISSED** [review]

File: packages/opencode/test/altimate/workspace/browser-handoff.test.ts [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+// ─────────────────────────────────────────────────────────────────────────────
+// End-to-end via runHandoffWithOpener — real loopback, injected browser-open
+// ─────────────────────────────────────────────────────────────────────────────
+
+describe("runHandoffWithOpener end-to-end", () => {
+  beforeEach(() => stubCreds("acme", "https://api.myaltimate.com"))
+  afterEach(() => unstubCreds())
+
+  test("happy path: valid callback resolves with workspaceId + tenant", async () => {
+    const result = await runHandoffWithOpener(
```
[the remark this answers is not in the document]
Comment (sahrizvi) [reply packages/opencode/test/altimate/workspace/browser-handoff.test.ts:145]: Fixed in [`23678b39`](https://github.com/AltimateAI/altimate-code/pull/1100/commits/23678b3928) — added a dedicated CSRF-mismatch test inside the `runHandoffWithOpener end-to-end` describe block. Fires a rogue callback with a wrong-state guess first (`workspace_id=999`), then a legitimate callback with the real state (`workspace_id=1`), and asserts the legitimate one wins — the rogue `999` never resolves the promise.

File: packages/opencode/src/altimate/workspace/browser-handoff.ts [hunk 2]
```diff
@@ -0,0 +1,533 @@
...
+/** Start a per-flow loopback listener on the first available port in the
+ * shared 7317..7325 range. Own server, own pending map — no coupling to the
+ * OAuth listener in altimate.ts. */
+async function startListener(pending: HandoffPending): Promise<{ server: Server; port: number }> {
+  const server = createServer((req, res) => {
+    const port = (server.address() as { port?: number } | null)?.port ?? CALLBACK_PORT_MIN
+    const url = new URL(req.url || "/", `http://127.0.0.1:${port}`)
+    if (url.pathname !== "/workspace-bound") {
+      res.writeHead(404)
+      res.end("Not found")
```
[the remark this answers is not in the document]
Comment (sahrizvi) [reply packages/opencode/src/altimate/workspace/browser-handoff.ts:220]: Fixed in [`b5273f31`](https://github.com/AltimateAI/altimate-code/pull/1100/commits/b5273f310e) — added the DNS-rebinding `Host` guard at the top of the handler in `browser-handoff.ts`. Rejects anything except `127.0.0.1:<port>` and `localhost:<port>` with 400, before state validation runs.

File: packages/opencode/src/altimate/workspace/browser-handoff.ts [hunk 3]
```diff
@@ -0,0 +1,533 @@
...
+  let settled = false
+  const closeListener = () => {
+    if (listenerHandle) {
+      try {
+        listenerHandle.server.close()
+      } catch {
+        /* best effort */
+      }
+      listenerHandle = undefined
+    }
```
[the remark this answers is not in the document]
Comment (sahrizvi) [reply packages/opencode/src/altimate/workspace/browser-handoff.ts:446]: Fixed in [`b5273f31`](https://github.com/AltimateAI/altimate-code/pull/1100/commits/b5273f310e) — followed both `server.close()` sites (the `closeListener()` in the pending-promise path and the port-exhaustion drop at line ~348) with `server.closeAllConnections?.()`. Optional-call guard keeps it safe on Node <18.2.

File: packages/opencode/src/plugin/tui/altimate/workspace-sidebar.tsx [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+          <>
+            <text fg={theme().textMuted}>
+              {b().datamateName}
+              <Show when={getResolvedWorkspaceId() === b().datamateId}>
+                {" (pinned via --workspace)"}
+              </Show>
+            </text>
+            <Show when={manageUrl()}>
+              {(u) => <text fg={theme().textMuted}>{u()}</text>}
+            </Show>
```
[the remark this answers is not in the document]
Comment (sahrizvi) [reply packages/opencode/src/plugin/tui/altimate/workspace-sidebar.tsx:121]: Addressed in [`b5273f31`](https://github.com/AltimateAI/altimate-code/pull/1100/commits/b5273f310e) — went with option (b) of the review: added a code comment on the sidebar spelling out what "pinned via --workspace" actually means ("this SESSION was launched with --workspace and it resolved to this id", not "current binding was set by --workspace"). `getResolvedWorkspaceId()` already encodes "was passed AND resolved" at the env-var level (`launch-resolve.ts` bails on `undefined` before setting it), so the pin never falsely appears for a session that wasn't launched with the flag. The relink-to-same-id case remains as documented imprecision.

File: packages/opencode/src/altimate/workspace/browser-handoff.ts [hunk 5]
```diff
@@ -0,0 +1,533 @@
...
+    try {
+      await new Promise<void>((resolve, reject) => {
+        const onErr = (err: NodeJS.ErrnoException) => reject(err)
+        server.once("error", onErr)
+        server.listen(port, "127.0.0.1", () => {
+          server.removeListener("error", onErr)
+          resolve()
+        })
+      })
+      // Post-listen persistent error handler. Without this, any socket-level
```
[the remark this answers is not in the document]
Comment (sahrizvi) [reply packages/opencode/src/altimate/workspace/browser-handoff.ts:330]: Confirmed — keeping the explicit `"127.0.0.1"` bind. The DNS-rebinding follow-up in [`b5273f31`](https://github.com/AltimateAI/altimate-code/pull/1100/commits/b5273f310e) added the `Host`-header guard on top of it, so the loopback bind + Host validation form the two-layer defense.

File: packages/opencode/src/altimate/workspace/browser-handoff.ts [hunk 6]
```diff
@@ -0,0 +1,533 @@
...
+      message: err instanceof Error ? err.message : String(err),
+    }
+  }
+
+  const state = randomBytes(16).toString("hex")
+
+  // Register pending, then bind the listener. Timeout owns rejection with
+  // reason "timeout"; the listener's own reject paths mark their own reasons.
+  let listenerHandle: { server: Server; port: number } | undefined
+  // ``settled`` reflects whether ``pending.resolve``/``pending.reject`` has
```
[the remark this answers is not in the document]
Comment (sahrizvi) [reply packages/opencode/src/altimate/workspace/browser-handoff.ts:426]: Confirmed — keeping `randomBytes(16).toString("hex")` as the CSRF-state source. The state comparison in the callback handler runs in constant time after the Host-header check ([`b5273f31`](https://github.com/AltimateAI/altimate-code/pull/1100/commits/b5273f310e)).

File: packages/opencode/src/cli/cmd/link.ts [hunk 7]
```diff
@@ -161,12 +200,124 @@ export const LinkCommand = cmd({
...
+  bindSpin.start("Linking workspace...")
+  try {
+    const res = await WorkspaceApi.bindExisting(result.workspaceId, identifier)
+    await recordApprovedBinding(directory, {
+      datamateId: res.binding.datamate_id,
+      datamateName: res.binding.datamate_name,
+      repoRemote: res.binding.repo_remote,
+      projectPath: res.binding.project_path,
+      linkedAt: Date.now(),
+    })
```
[the remark this answers is not in the document]
Comment (sahrizvi) [reply packages/opencode/src/cli/cmd/link.ts]: Fixed in [`d011067f`](https://github.com/AltimateAI/altimate-code/pull/1100/commits/d011067fec) — the browser-flow `recordApprovedBinding` call now uses `identifier.projectPath ?? directory`, matching the two other call sites in the same file. A `-d ./myproj` invocation now caches under the same key as `altimate-code link` from the project root.

File: packages/opencode/src/altimate/workspace/browser-handoff.ts [hunk 8]
```diff
@@ -0,0 +1,557 @@
...
+    // callback URL carrying those forms would slip past ``isInteger`` and
+    // reach the bind payload. Requiring a plain decimal-digit string first
+    // is the tight gate. (cubic cycle 4/5.)
+    if (!/^[1-9][0-9]*$/.test(workspaceIdRaw)) {
+      const msg = `Invalid workspace_id: ${workspaceIdRaw}`
+      respond(400, htmlError(msg))
+      pending.reject(markReason(new Error(msg), "error"))
+      return
+    }
+    const workspaceId = Number(workspaceIdRaw)
```
[the remark this answers is not in the document]
Comment (sahrizvi) [reply packages/opencode/src/altimate/workspace/browser-handoff.ts:285]: Fixed in [`d011067f`](https://github.com/AltimateAI/altimate-code/pull/1100/commits/d011067fec) — swapped `Number.isInteger` for `Number.isSafeInteger`. Still unreachable in practice today, but the one-word guard closes the drift window if IDs ever grow past 2^53.

File: packages/opencode/src/altimate/workspace/state.ts [hunk 9]
```diff
@@ -109,6 +91,32 @@ function readCache(): CacheFile | null {
...
+ * dead code — every subsequent read hits the direct key lookup. */
+function migrateToCanonicalKeys(cache: CacheFile): CacheFile {
+  const migrated: Record<string, CachedBinding> = {}
+  for (const [k, v] of Object.entries(cache.bindings)) {
+    const canon = canonicalizeKey(k)
+    const existing = migrated[canon]
+    if (!existing || existing.linkedAt <= v.linkedAt) migrated[canon] = v
+  }
+  const next: CacheFile = { ...cache, bindings: migrated }
+  writeCache(next)
```
[the remark this answers is not in the document]
Comment (sahrizvi) [reply packages/opencode/src/altimate/workspace/state.ts]: Fixed in [`d011067f`](https://github.com/AltimateAI/altimate-code/pull/1100/commits/d011067fec) — wrapped the `writeCache(next)` inside `migrateToCanonicalKeys` in try/catch. The migrated shape is still returned in-memory for the current `readLocalBinding` call; the next successful `recordApprovedBinding` persists the canonical form. A read-only state dir no longer surfaces "Workspace setup failed" on the offline-fallback path.

File: packages/opencode/src/plugin/tui/altimate/workspace.tsx [hunk 10]
```diff
@@ -183,6 +228,249 @@ function OfferDialog(props: OfferProps) {
...
+  api.ui.dialog.clear()
+  api.ui.toast({
+    variant: "info",
+    message: "Opening browser to set up your workspace — approve there, then check back here for the confirmation.",
+  })
+  // Supersede any still-open handoff before starting a new one.
+  if (activeHandoffAbort) activeHandoffAbort.abort()
+  activeHandoffAbort = new AbortController()
+  const signal = activeHandoffAbort.signal
+  const result: HandoffResult = await openWorkspaceBrowserHandoff({ identifier, projectName, signal })
```
[the remark this answers is not in the document]
Comment (sahrizvi) [reply packages/opencode/src/plugin/tui/altimate/workspace.tsx:350]: Fixed in [`85161a58`](https://github.com/AltimateAI/altimate-code/pull/1100/commits/85161a5824) — early-return in `runBrowserHandoff` when `result.reason === "aborted"`. The newer flow's "Opening browser..." toast is the correct signal; the superseded promise exits silently instead of hitting `toastHandoffFailure`'s default red toast. Regression from my own AbortController fix — thanks for catching.

**Review by saravmajestic — APPROVED** [review]
